src/read/pipelining.rs

```diff
@@ -0,0 +1,1209 @@
+        },
+        pipe::{unix::create_pipe, ReadSplicer, WriteSplicer},
+        util::{copy_via_buf, TakeWrite},
+    };
+
+    use super::{
+        handle_creation::{
+            transform_entries_to_allocated_handles, AllocatedHandles, HandleCreationError,
+            ZipDataHandle,
+        },
+        path_splitting::{lexicographic_entry_trie, PathSplitError},
+    };
+
+    /// Errors encountered during the split pipelined extraction process.
+    #[derive(Debug, Display, Error)]
+    pub enum SplitExtractionError {
+        /// i/o error: {0}
+        Io(#[from] io::Error),
+        /// zip error: {0}
+        Zip(#[from] ZipError),
+        /// path split error: {0}
+        PathSplit(String),
+        /// handle creation error: {0}
+        HandleCreation(#[from] HandleCreationError),
+    }
+
+    impl<'a> From<PathSplitError<'a>> for SplitExtractionError {
+        fn from(e: PathSplitError<'a>) -> Self {
+            let msg = format!("{}", e);
+            Self::PathSplit(msg)
+        }
+    }
+
+    /* TODO: make this share code with find_data_start()! */
+    fn get_or_find_data_start<InF>(data: &ZipFileData, input_file: InF) -> Result<u64, ZipError>
+    where
+        InF: InputFile,
+    {
+        // TODO: use .get_or_try_init() once stabilized to provide a closure returning a Result!
+        if let Some(data_start) = data.data_start.get() {
+            return Ok(*data_start);
+        }
+
+        let block = {
+            let block: MaybeUninit<[u8; mem::size_of::<ZipLocalEntryBlock>()]> =
+                MaybeUninit::uninit();
+            let mut block: [MaybeUninit<u8>; mem::size_of::<ZipLocalEntryBlock>()] =
+                unsafe { mem::transmute(block) };
+
+            input_file.pread_all(data.header_start, &mut block[..])?;
+
+            let block: MaybeUninit<[u8; mem::size_of::<ZipLocalEntryBlock>()]> =
+                unsafe { mem::transmute(block) };
+            unsafe { block.assume_init() }
+        };
+        // Parse static-sized fields and check the magic value.
+        let block = ZipLocalEntryBlock::interpret(block.as_ref())?;
+
+        // Calculate the end of the local header from the fields we just parsed.
+        let variable_fields_len: u64 =
+            // Each of these fields must be converted to u64 before adding, as the result may
+            // easily overflow a u16.
+            block.file_name_length as u64 + block.extra_field_length as u64;
+        let local_entry_block_size: u64 = mem::size_of::<ZipLocalEntryBlock>().try_into().unwrap();
+        let data_start: u64 = data.header_start + local_entry_block_size + variable_fields_len;
+
+        // Set the value so we don't have to read it again.
+        match data.data_start.set(data_start) {
+            Ok(()) => (),
+            // If the value was already set in the meantime, ensure it matches.
+            Err(_) => {
+                assert_eq!(*data.data_start.get().unwrap(), data_start);
```

Comment:
Change this to `debug_assert_eq!` for performance reasons.

Reply:
Done!